### PR TITLE
[Failure store] Refactoring data stream lifecycle to facilitate managing failure store.

### PR DESCRIPTION
**Pull request by gmarouli**
In this PR we refactor data stream and data stream lifecycle related code. The goal of this small refactoring is to make it easier to extend DSL to manage failure store.

- The retrieval of backing indices past the retention period. 
- The rollover execution during the DSL run to handle more than one write index.

### CODE REVIEW

File: modules/data-streams/src/main/java/org/elasticsearch/datastreams/lifecycle/DataStreamLifecycleService.java [hunk 1]
```diff
@@ -343,35 +343,17 @@ void run(ClusterState state) {
...
             Set<Index> indicesToExcludeForRemainingRun = new HashSet<>();
-            // the following indices should not be considered for the remainder of this service run:
-            // 1) the write index as it's still getting writes and we'll have to roll it over when the conditions are met
-            // 2) tsds indices that are still within their time bounds (i.e. now < time_series.end_time) - we don't want these indices to be
+
+            // This is the pre-rollover write index. It may or may not be the write index after maybeExecuteRollover has executed,
+            // depending on rollover criteria.
+            // The write index is excluded from the remaining run, as it's still getting writes and we'll have to roll it over
+            // when the conditions are met.
+            indicesToExcludeForRemainingRun.addAll(maybeExecuteRollover(state, dataStream));
```
Comment (gmarouli) [inline modules/data-streams/src/main/java/org/elasticsearch/datastreams/lifecycle/DataStreamLifecycleService.java:353]: Changing `maybeExecuteRollover` in this way allows us to return both the backing write index and the failure store write index and exclude them from the next steps.

File: server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java [hunk 2]
```diff
@@ -822,32 +822,39 @@ public List<Round> getDownsamplingRoundsFor(
...
-                if (nowMillis >= indexLifecycleDate.getMillis() + age.getMillis()) {
-                    if (indicesPredicate == null || indicesPredicate.test(indexMetadata)) {
-                        olderIndices.add(index);
-                    }
-                }
+            if (isIndexOderThan(index, retentionPeriod.getMillis(), nowSupplier.getAsLong(), indicesPredicate, indexMetadataSupplier)) {
+                olderIndices.add(index);
             }
         }
         return olderIndices;
```
Comment (nielsbauman) [inline server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:836]: Nit: I usually think the `Stream` API is a bit more elegant. I don't feel strong enough about it to actually "request changes", but thought I'd mention it. AKA, up to you :).

**Review by nielsbauman — APPROVED** [review]
LGTM, just left two minor comments. Thanks for splitting this up into multiple/smaller PRs <3

File: server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java (hunk 2 above)
Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:836]: Thanks Niels, I agree, I was just trying to balance performance and elegance. Let me explain, I think in the stream approach we create:
- Here we create one list in which we add elements and we use two iterators.
- In the stream approach, we would first need to concatenate the two lists into a new list, or convert them into streams and then concatenate these streams. Both felt a bit too much overhead.

But I do not feel too strongly about it. Happy to revert back to streams :). What do you think?

Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:836]: Ahh I hadn't realized that you're going to add the failure store indices afterwards as well of course. Maybe, when you actually implement that part, I'd play around with having a local `Consumer<Index>` variable that you can then use to `indices.forEach(consumer)` and `failureStoreIndices.forEach(consumer)`. But not sure if that ends up being (more) elegant or not. I definitely agree with the setup you have now so I'll resolve this comment, thanks :)

Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:836]: Interesting idea, let's work on it on the follow up :)

**Comment by gmarouli** [comment]
@elasticmachine update branch

**Review by jbaiera — COMMENTED** [review]
LGTM post merge, though I spotted a typo that might be worth fixing in the next round 😄

**Pull request merged**